src/Builder.jl

```diff
@@ -76,33 +76,101 @@ Selectors.strict{T <: DocumentPipeline}(::Type{T}) = false
 
 function Selectors.runner(::Type{SetupBuildDirectory}, doc::Documents.Document)
     Utilities.log(doc, "setting up build directory.")
+
     # Frequently used fields.
     build  = doc.user.build
     source = doc.user.source
 
+    # The .user.source directory must exist.
+    isdir(source) || error("source directory '$(abspath(source))' is missing.")
+
+    # We create the .user.build directory.
+    # If .user.clean is set, we first clean the existing directory.
     doc.user.clean && isdir(build) && rm(build; recursive = true)
     isdir(build) || mkdir(build)
-    if isdir(source)
-        for (root, dirs, files) in walkdir(source)
-            for dir in dirs
-                d = normpath(joinpath(build, relpath(root, source), dir))
-                isdir(d) || mkdir(d)
-            end
-            for file in files
+
+    # We'll walk over all the files in the .user.source directory.
+    # The directory structure is copied over to .user.build. All files, with
+    # the exception of markdown files (identified by the extension) are copied
+    # over as well, since they're assumed to be images, data files etc.
+    # Markdown files, however, are stored in a list, to be used later.
+    mdpages = Compat.UTF8String[]
+    for (root, dirs, files) in walkdir(source)
+        for dir in dirs
+            d = normpath(joinpath(build, relpath(root, source), dir))
+            isdir(d) || mkdir(d)
+        end
+        for file in files
+            if endswith(file, ".md")
+                # pushing the path of the file relative to source into mdpages
+                push!(mdpages, Utilities.srcpath(source, root, file))
+            else
                 src = normpath(joinpath(root, file))
                 dst = normpath(joinpath(build, relpath(root, source), file))
-                # Non-markdown files are simply copied over to `build`.
-                # Markdown files get added to the document tree as `Page` objects.
-                endswith(src, ".md") ?
-                    Documents.addpage!(doc, src, dst) :
-                    cp(src, dst; remove_destination = true)
+                cp(src, dst; remove_destination = true)
             end
         end
-    else
-        error("source directory '$(abspath(source))' is missing.")
+    end
+
+    # Load all the pages to the document.
+    for path in mdpages
+        src = joinpath(source, path)
+        dst = joinpath(build, path)
+        Documents.addpage!(doc, src, dst)
```

Comment:
Both would work, but I like that this breaks it into two separate steps -- walking the file system and actually parsing the files. The copying should be moved out by that argument too, true, but that should be moved into the writers anyway at some point (with the walk just creating a list of files).